### ISSUE TITLE
Variant syntax of optional arguments

### ISSUE THREAD

**Issue by fujidana** [fujidana/vscode-igorpro#27]
The extension recently supported the syntax for optional parameters (`displayhelptopic "Optional Parameters"`), after a bug reporting in Issue #5 (Optional arguments are not correctly parsed). However, after that I found that the extension fails to parse several IPF files bundled with Igor Pro and that a variant expression of optional-argument syntax is used in those codes.

```igorpro
// usage in the official manual
Function Example(Variable a, [ Variable b, double c ])

// code in /Applications/Igor Pro 9 Folder/WaveMetrics Procedures/Analysis/Bivariate Histogram 2.ipf (macOS version)
Function BivariateHist(XWave, YWave, XBins, YBins, histWaveName [, doClassWave])
```

The difference is that a comma is inside or outside the brackets. While the latter form is probably not in official documents, Igor Pro can compile it without any error. Considering that items inside the brackets represents what can be omitted, the latter expression looks reasonable.

Next I investigated the cases of several variants I can imagine and found that, surprisingly, all the following are acceptable:

```igorpro
Function test(a, [b, c]) // documented form
Function test(a[, b, c]) // not documented but reasonable; bundled codes actually use this
Function test(a, [, b, c]) // combination of the first two, verbose.
Function test(a [b, c]) // no comma between required and optional, strange (look like wave access)
Function test([b, c]) // natural
Function test([, b, c]) // verbose
Function test(,[b, c]) // verbose
Function test(,[, b, c]) // very verbose
```

In addition, just one trailing comma after the last parameter can be accepted (similar to JavaScript?), that is, `c` in the example above can be replaced with `c,`, e.g., ` test(a, [b, c,])`. In cases of functions without optional parameters, `test(a,)` and `test(,)` are acceptable patterns. This trailing comma must be zero or one; two or more commas, e.g.,`test(a, [b, c,,])` causes a compile error.

The current rule of the extension's parser is a bit strict and only the documented form of optional-parameter syntax is supported. I will modify it more leniently.

**Comment by fujidana** [comment]
Considering that the preceding and trailing extra commas are prohibited in some other places, I think extra commas should not be used in parameters, while current built-in compiler allows them in a parameter list. 

The updated parser will regard the following expressions as regular forms and will report an error or a warning if extra commas are used.

```igorpro
// with required parameters
Function test(a, [b, c]) // form that appears in the official doc
Function test(a[, b, c]) // variant.

// without required parameter
Function test([b, c])
```

**Comment by fujidana** [comment]
I changed the AST structure in order to take measures against #5 and this issue but during that I missed to update the source code related to code-hint part that depends on the AST. Because of this, a bug such that function-parameter part of code-hint (e.g., hover) is not properly shown was introduced. Now I fixed it in f942c11 and 0c2b265.
